### PR TITLE
Switch to using the REST client when talking to Elasticsearch

### LINKED ISSUE #402: API Reliability: DNS on "Node not found" failure
Currently Elastic Cloud / AWS will switch out the ip address of ES nodes which causes a problem for consumers using the ES Transport client (in our case) as the IP address is resolved on startup and never again, resulting in occasional outages.

We should fix it :)

### PR DESCRIPTION
**Pull request by alexwlchan**
### What is this PR trying to achieve?

We currently use the transport client for talking to Elasticsearch, but this is causing flakiness. Specifically, it seems to resolve the host to an IP address at startup, then continues to use that address even if the DNS records change. Using the REST client means we check the DNS records every time. This is also the recommended approach by Elasticsearch.

Resolves #402.

Todo:

- [x] How do we pass authentication info to the REST client?
- [x] Strip out authentication info for X-Pack/the transport client
- [x] Make sure the ingestor still works
- [x] Make sure the API still works

### Who is this change for?

Users of our API who don’t want it to be flakey.

### Have the following been considered/are they needed?

- [ ] Deployed new versions
- [ ] Run `terraform apply`.

### CODE REVIEW

File: api/src/main/scala/uk/ac/wellcome/platform/api/models/DisplayWork.scala [hunk 1]
```diff
@@ -1,8 +1,9 @@
 package uk.ac.wellcome.platform.api.models
 
 import com.fasterxml.jackson.annotation.JsonProperty
-import com.sksamuel.elastic4s.searches.RichSearchHit
-import org.elasticsearch.action.get.GetResponse
+
```
Comment (alexwlchan) [inline api/src/main/scala/uk/ac/wellcome/platform/api/models/DisplayWork.scala]: drop the blank line

File: common/src/main/scala/uk/ac/wellcome/finatra/modules/ElasticClientModule.scala [hunk 2]
```diff
@@ -3,11 +3,19 @@ package uk.ac.wellcome.finatra.modules
...
 import com.twitter.inject.TwitterModule
 import org.elasticsearch.common.settings.Settings
 
+import org.apache.http.auth.AuthScope
+import org.apache.http.impl.nio.client.HttpAsyncClientBuilder
+import org.apache.http.auth.UsernamePasswordCredentials
+import org.apache.http.impl.client.BasicCredentialsProvider
+import org.apache.http.HttpHost
+import org.elasticsearch.client.RestClient
+import org.elasticsearch.client.RestClientBuilder.HttpClientConfigCallback
```
Comment (alexwlchan) [inline common/src/main/scala/uk/ac/wellcome/finatra/modules/ElasticClientModule.scala]: tidy these up

File: common/src/main/scala/uk/ac/wellcome/finatra/modules/ElasticClientModule.scala [hunk 3]
```diff
@@ -28,43 +36,59 @@ object XPackConfigModule extends TwitterModule {
...
+    // //     (defaults, config) =>
+    // //       defaults
+    // //         .put("xpack.security.transport.ssl.enabled", config.ssl)
+    // //         .put("request.headers.X-Found-Cluster", clusterName())
+    // //         .put("xpack.security.user", config.user))
+    // //   .build()
+    //
+    // settings.getAsMap.asScala.foreach {
+    //   case (k, v) => info(s"ElasticClient setting: $k=$v")
+    // }
```
Comment (alexwlchan) [inline common/src/main/scala/uk/ac/wellcome/finatra/modules/ElasticClientModule.scala]: delete, don't comment out

File: common/src/test/scala/uk/ac/wellcome/elasticsearch/mappings/WorksIndexTest.scala [hunk 4]
```diff
@@ -77,17 +76,21 @@ class WorksIndexTest
...
+    //     .mappings
+    //     .get(indexName)
+    //     .get(itemType)
+    //     // .mappingFor(indexName / itemType)
+    //     // .getSourceAsMap
+    //     // .toMap
+    //   mappings("properties")
+    //     .asInstanceOf[util.Map[String, AnyRef]]
+    //     .keys should contain("work")
+    // }
```
Comment (alexwlchan) [inline common/src/test/scala/uk/ac/wellcome/elasticsearch/mappings/WorksIndexTest.scala]: this needs putting back

File: common/src/test/scala/uk/ac/wellcome/test/utils/ElasticSearchLocal.scala [hunk 5]
```diff
@@ -1,16 +1,28 @@
...
 import scala.concurrent.Future
 
+import uk.ac.wellcome.finatra.modules.ElasticCredentials
+import org.apache.http.auth.AuthScope
+import org.apache.http.impl.nio.client.HttpAsyncClientBuilder
+import org.apache.http.auth.UsernamePasswordCredentials
+import org.apache.http.impl.client.BasicCredentialsProvider
+import org.apache.http.HttpHost
+import org.elasticsearch.client.RestClient
+import org.elasticsearch.client.RestClientBuilder.HttpClientConfigCallback
```
Comment (alexwlchan) [inline common/src/test/scala/uk/ac/wellcome/test/utils/ElasticSearchLocal.scala]: these imports need fixing

File: common/src/test/scala/uk/ac/wellcome/test/utils/ElasticSearchLocal.scala [hunk 6]
```diff
@@ -22,12 +34,18 @@ trait ElasticSearchLocal
...
-  val elasticClient =
-    XPackElasticClient(settings, ElasticsearchClientUri("localhost", 9300))
+  val restClient = RestClient
+    .builder(new HttpHost("localhost", 9200, "http"))
+    .setHttpClientConfigCallback(new ElasticCredentials())
+    .build()
+
+  val elasticClient = HttpClient.fromRestClient(restClient)
+
+  // val elasticClient = HttpClient(ElasticsearchClientUri("localhost", 9200))
```
Comment (alexwlchan) [inline common/src/test/scala/uk/ac/wellcome/test/utils/ElasticSearchLocal.scala]: delete

File: ingestor/src/test/scala/uk/ac/wellcome/platform/ingestor/IngestorFeatureTest.scala [hunk 7]
```diff
@@ -51,7 +51,7 @@ class IngestorFeatureTest
     eventually {
       val hitsFuture = elasticClient
         .execute(search(s"$indexName/$itemType").matchAllQuery())
-        .map(_.hits)
+        .map { _.hits.hits }
```
Comment (alexwlchan) [inline ingestor/src/test/scala/uk/ac/wellcome/platform/ingestor/IngestorFeatureTest.scala]: revert

File: project/Dependencies.scala [hunk 8]
```diff
@@ -40,6 +40,7 @@ object Dependencies {
   )
   val esDependencies: Seq[ModuleID] = Seq(
     "com.sksamuel.elastic4s" %% "elastic4s-core" % versions.elastic4s,
+    "com.sksamuel.elastic4s" %% "elastic4s-http" % versions.elastic4s,
     "com.sksamuel.elastic4s" %% "elastic4s-testkit" % versions.elastic4s % "test",
     "com.sksamuel.elastic4s" %% "elastic4s-xpack-security" % versions.elastic4s
```
Comment (alexwlchan) [inline project/Dependencies.scala]: can probably lose this

File: common/src/main/scala/uk/ac/wellcome/finatra/modules/ElasticClientModule.scala [hunk 9]
```diff
@@ -28,43 +36,59 @@ object XPackConfigModule extends TwitterModule {
     }
 }
 
+class ElasticCredentials extends HttpClientConfigCallback {
+  val credentials = new UsernamePasswordCredentials("elastic", "changeme")
```
Comment (alexwlchan) [inline common/src/main/scala/uk/ac/wellcome/finatra/modules/ElasticClientModule.scala]: get these from flags instead

**Comment by alexwlchan** [comment]
Okay, I think this is ready for review.

File: common/src/main/scala/uk/ac/wellcome/finatra/modules/ElasticClientModule.scala [hunk 10]
```diff
@@ -3,68 +3,43 @@ package uk.ac.wellcome.finatra.modules
...
-    Option(xpackEnabled()).collect {
-      case true =>
-        XPackConfig(user(), sslEnabled())
-    }
+import org.apache.http.auth.{AuthScope, UsernamePasswordCredentials}
+import org.apache.http.HttpHost
+import org.apache.http.impl.client.BasicCredentialsProvider
+import org.apache.http.impl.nio.client.HttpAsyncClientBuilder
+
+class ElasticCredentials (username: String, password: String) extends TwitterModule with HttpClientConfigCallback {
```
Comment (alicefuzier) [inline common/src/main/scala/uk/ac/wellcome/finatra/modules/ElasticClientModule.scala]: does it need to extend `TwitterModule`?

Comment (alexwlchan) [inline common/src/main/scala/uk/ac/wellcome/finatra/modules/ElasticClientModule.scala]: No, good spot – an earlier version of this code defined the flags in this class, but then I moved them down to `ElasticCredentials`. I’ll remove it.

**Comment by alexwlchan** [comment]
Fixed the `TwitterModule` subclassing; will merge once tests pass.

**Pull request merged**